File: Python/0005_longest_palindromic_substring/0005_longest_palindromic_substring.py
class Solution:
    def longestPalindrome(self, s):
        """
        :type s: str
        :rtype: str
        """
        if len(s) == 0:
            return s
        dp = [[False for i in range(len(s))] for j in range(len(s))]
        start = 0
        maximum = 0
        for i in range(len(s)):
            for j in range(0, i + 1):
                if s[i] == s[j] and (i - j <= 2 or dp[j + 1][i - 1]):
                    dp[j][i] = True
                if dp[j][i] and maximum < i - j + 1:
                    maximum = i - j + 1
                    start = j
        return s[start:start + maximum]

File: Cplusplus/0005_longest_palindromic_substring/0005_longest_palindromic_substring.cpp
class Solution {
public:
    string longestPalindrome(string s) {
        if (s.size() == 0)
            return s;
        bool dp[1000][1000] = { false };
        int start = 0;
        int max = 0;
        string res;
        for (int i = 0; i < s.size(); i++) {
            for (int j = 0; j <= i; j++) {
                if (s.at(i) == s.at(j) && (i - j <= 2 || dp[j + 1][i - 1])) {
                    dp[j][i] = true;
                }
                if (dp[j][i] && max < i - j + 1) {
                    max = i - j + 1;
                    start = j;
                }
            }
        }
        for (int i = start; i < start + max; i++) {
            res += s[i];
        }
        return res;
    }
};